src/main/java/org/apache/sysds/runtime/matrix/data/LibMatrixIMGTransform.java

```diff
@@ -0,0 +1,232 @@
+import org.apache.sysds.runtime.functionobjects.CTable;
+import org.apache.sysds.runtime.functionobjects.SwapIndex;
+import org.apache.sysds.runtime.instructions.InstructionUtils;
+import org.apache.sysds.runtime.matrix.operators.*;
+import org.apache.sysds.runtime.util.DataConverter;
+
+import static org.apache.commons.math3.util.FastMath.floor;
+
+/**Separate class for generating the transformation matrix based on an affine transformation matrix and a matrix
+ * containing the dimensions of the input and the target dimensions of the output
+ */
+public class LibMatrixIMGTransform {
+
+    protected static final Log LOG = LogFactory.getLog(LibMatrixIMGTransform.class.getName());
+
+
+//     * affine transformation matrix for calculated for a picture of original size and target dimensions
+//     * see: https://en.wikipedia.org/wiki/Affine_transformation
+//     *
+//     * #orig_w = as.scalar(dimMat[1,2])
+//     * #orig_h = as.scalar(dimMat[1,1])
+//     * #out_w = as.scalar(dimMat[2,2])
+//     * #out_h = as.scalar(dimMat[2,1])
+//     * T_inv = inv(transMat)
+//     *
+//     * ## coordinates of output pixel-centers linearized in row-major order
+//     * coords = matrix(1, rows=3, cols=out_w*out_h)
+//     * coords[1,] = t((seq(0, out_w*out_h-1) %% out_w) + 0.5)
+//     * coords[2,] = t((seq(0, out_w*out_h-1) %/% out_w) + 0.5)
+//     * # compute sampling pixel indices
+//     * coords = floor(T_inv %*% coords) + 1
+//     * inx = t(coords[1,])
+//     * iny = t(coords[2,])
+//     * # any out-of-range pixels, if present, correspond to an extra pixel with fill_value at the end of the input
+//     * index_vector = (orig_w *(iny-1) + inx) * ((0<inx) & (inx<=orig_w) & (0<iny) & (iny<=orig_h))
+//     * index_vector = t(index_vector)
+//     * xs = ((index_vector == 0)*(orig_w*orig_h +1)) + index_vector
+//     * #if(min(index_vector) == 0){
+//     * #  ys=cbind(img_in, matrix(fill_value,nrow(img_in), 1))
+//     * #}else{
+//     * #  ys = img_in
+//     * #}
+//     * ind= matrix(seq(1,ncol(xs),1),1,ncol(xs))
+//     * z = table(xs, ind)
+//     * zMat = transMat
+//     * isFillable = as.double(min(index_vector) == 0)
+    /** This method produces a transformation matrix for affine transformation (see Wikipedia: "Affine transformation").
+     * It takes a transformation 3x3 matrix (last row is always 0,0,1 for images), called affine matrix, and a dimension matrix, where the
+     * original image dimensions (width x height) are stored alongside the output image dimensions (width x height)
+     * @param transMat affine 3x3 matrix for image transformations
+     * @param dimMat 2x2 matrix with original and output image dimensions
+     * @param threads number of threads for use in different methods
+     *
+     * @return array of two matrix blocks, 1st is the transformation matrix, 2nd a 1x1 matrix with 1 or 0
+     */
+    public static MatrixBlock[] transformationMatrix(MatrixBlock transMat, MatrixBlock dimMat, int threads) {
```

Comment:
Overall, it is a good starting point for the operation. 
You can probably benefit from some of the instructions to use threads, and some of the scalar operations where the MatrixBlock is allocated can probably be hand optimized. 

It would be good if you could modify the lines, to not have the comments on the lines but on the line above.

Reply:
Ill try to find some possible improvements, but i am slighly limited on time before the presentation. Lines are modified